doc/source/seedbox-config.rst
Sample Configuration
====================
::

        [DEFAULT]

        # Base path (string value)
        #base_path = /home/username

        # Location torrent client stores data files (string value)
        #base_client_path = $base_path/torrents/data


        [database]

        # The connection string used to connect to the database (string value)
        #connection = sqlite:///$config_dir/torrent.db

        # Timeout before idle sql connections are reaped (integer value)
        #idle_timeout = 3600

        # Verbosity of SQL debugging information. 0=None, 100=All (integer
        # value)
        #connection_debug = 0


        [process]

        # max processes to use for performing sync of torrents (integer value)
        #max_processes = 4

        # name of tasks associated with prepare phase (list value)
        #prepare = filecopy, fileunrar

        # name of tasks associated with activate phase (list value)
        #activate = filesync

        # name of tasks associated with complete phase (list value)
        #complete = filedelete


        [tasks]

        # Location to temp media copies for syncing to library (string value)
        #sync_path = /tmp/sync


        [tasks_filesync]

        # rsync dryrun option (boolean value)
        #dryrun = false

        # rsync verbose option (boolean value)
        #verbose = false

        # rsync progress option (boolean value)
        #progress = false

        # rsync perms option (boolean value)
        #perms = true

        # rsync delayupdates option (boolean value)
        #delayupdates = true

        # rsync recursive option (boolean value)
        #recursive = true

        # rsync chmod option (string value)
        #chmod = ugo+rwx

        # rsync-ssh identity option (ssh key) (string value)
        #identity = /home/username/.ssh/my_home.key

        # rsync-ssh port (string value)
        #port = 22

        # User name on remote system (ssh) (string value)
        #remote_user = username

        # Host name/IP Address of remote system (string value)
        #remote_host = home.example.com

        # rsync destination path (string value)
        #remote_path = /media/downloads


        [tasks_synclog]

        # Output directory for stdout files (string value)
        #stdout_dir = $config_dir/sync_out

        # Output directory for stderr files (string value)
        #stderr_dir = $config_dir/sync_err

        # Write output to stdout (boolean value)
        #stdout_verbose = false

        # Output verbose details about exceptions (boolean value)
        #stderr_verbose = true


        [torrent]

        # Location of the .torrent files (string value)
        #torrent_path = /home/username/.config/state

        # Location(s) of the media files (list value)
        #media_paths = $base_client_path/completed

        # Location of the downloading torrents (string value)
        #incomplete_path = $base_client_path/inprogress

        # List of video filetypes to support. (ignore others) (list value)
        #video_filetypes = .avi,.mp4,.mkv,.mpg

        # List of compressed filetypes to support. (ignore others) (list
        # value)
        #compressed_filetypes = .rar

        # Minimum file size of a media file (integer value)
        #minimum_file_size = 75000000


doc/source/api/seedbox.options.rst
The :mod:`seedbox.options` Module
=================================

.. automodule:: seedbox.options
  :members:
  :undoc-members:
  :show-inheritance:


ChangeLog
Changelog
=========

2.3.16
------

* [RELEASE] Update to version v2.3.16
* TASK] Auto image build on DockerHub and Quay
* [TASK] Updated cli and config file options doc
* [TASK] Refreshed documentation
* [TASK] Add development artifacts
* removed old release bash script

2.3.15
------

* [RELEASE] Update to version v2.3.15
* [TASK] Leverage invoke for managing releases
* [TASK] Simplified custom torrent parser
* [TASK] Update to latest pbr

2.3.14
------

* [RELEASE] Update to version v2.3.14
* [TASK] Database API access clean up
* [TASK] Drop support for Python 2.6 and 3.3
* [TASK] Cleanup options define and loading
* updated how source code documentation handled
* updated landscape config
* added landscape configuration
* [TASK] Basic cleanup of code base
* update doc formatting
* updated alpine version
* added .dockerignore
* added Dockerfile
* fix file permissions

2.3.13
------

* [RELEASE] Update to version v2.3.13
* [FEATURE] Update requirements
* updated version from 3 to 4
* [BUGFIX] Fix total_time column type

2.3.12
------

* [RELEASE] Update to version v2.3.12
* [TASK] Switch to oslo_config
* [FEATURE] Added Admin UI and REST APIs
* [TASK] Fixed error message formatting
* [TASK] Standards alignment with hacking
* [TASK] Requirements updates
* [TASK] Fixed error message formatting
* [TASK] Standards alignment with hacking
* [TASK] Requirements updates
* [TASK] Document updates
* [TASK] Remove intersphinx dependency
* [TASK] Upgrade version of oslo.config
* [TASK] Update README to include Waffle
* [TASK] Reduce output from update_version
* [TASK] General cleanup activities

2.3.11
------

* Update to version v2.3.11
* [TASK] General cleanup activities
* [TASK] Apply shell lint changes to update_version
* [TASK] Fix up spelling errors
* [BUGFIX] Fix capture of filename when coming from compressed file
* remove patch for what was oslo.config a bug; replaced original way of setting config_dir with set_default. Consistent response regardless of version of oslo.config
* minor change to handle the fact we do not publish the release branch to the remote, and git flow raises an exception when attempting to remove the remote branch. now catching that exception only and allowing script to continue with upload if everything else was successful

2.3.10
------

* Update to version v2.3.10
* updated release process to separate the packaging from the release via twine
* changed url from pypi to github

2.3.9
-----

* Update to version v2.3.9
* fixed minor pep8 issues within new test cases
* added more test cases
* update ChangeLog based on recent changes
* added logic to stop workflow from continuing execution when a task associated with a media file failed to further support retry
* code cleanup and simplification and duplication
* skip adding files found with the torrent if not found on filesystem to avoid processing a torrent w/o files to process; general cleanup and simplification of code
* replaced empty abstract methods with raise NotImplementedError for clarity and coverage
* converted from previous use of 1 or 0 to represent True and False when setting values to be stored in the database
* clean up to further simplify implementation of pid lockfile
* added links to requires.io
* added ignore raise NotImplementedError
* adding travis_retry during install travis config
* trying a different travis config
* added design documentation

2.3.8
-----

* Update to version v2.3.8
* updated to make sure subprocessext is patched for testing filesync
* removed log entry that resulted in breaking test code because it mocks part that checks for called once
* fixed testing class impacted by bug in oslo.config placeholder substitution
* fixed bug where execute method was not called such that rsync never happened; fixed bug where config dir was not found when defaulting to path including configuration directory
* fixed logging entries by defining logger for parser to avoid going to root logger
* fix issue where database file is not able to be opened when defaulting to configuration directory

2.3.7
-----

* Update to version v2.3.7
* fixed format of message; py26
* updated to support py26
* applied updates to support py26
* updated test cases to cover usage scenarios; added new test class to support generating sample config
* removed files from coverage that are entry point executables or copied from another library to provide coverage based on actual code from this code base
* fixed pep8 issue
* removed unnecessary docstrings in empty files
* fix ChangeLog by fixing which commit was associated with tag 2.2.4

2.3.6
-----

* Update to version v2.3.6
* test_save still causing an issue; commenting a line for now until periodic issue can be resolved

2.3.5
-----

* Update to version v2.3.5
* removed custom configuration generation code base and script
* fixed db save that fails randomly when two instances with different values not successful when asserting not equal
* fix pep8 issues
* add generation of sample config via oslo.config instead of custom version copied from oslo.incubator project
* undo travis configuration
* updating travis configuration
* removing copy of config fixture; leverage instance within oslo.config
* replaced extra imports for the purpose of making config options available; leverage library approach for importing options from other modules
* moved to using mockpatch from oslotest instead of local one
* simplified testing dependencies by leveraging oslotest which handles the setting up the base test class

2.3.4
-----

* Update to version v2.3.4
* removed unnecessary assignment of instance to variable; corrected docstring
* updated parameters to match actual abstract
* fixed TODO docstring
* updated year to 2014
* adding fetch origin as part of finish release

2.3.3
-----

* Update to version v2.3.3
* updated to tox 1.7.2 to fix posarg issue so updated tox.ini to use setup.py directly and allowed for removal of run_test.sh

seedbox/options.py
"""Handles configuration options.

Defines all the common configurations for the application, and then
manages loading all Options for system.
"""
import os

from oslo_config import cfg

CLI_OPTS = [
    cfg.BoolOpt('cron',
                default=False,
                help='Disable console output when running via cron'),
    cfg.StrOpt('logfile',
               metavar='LOG_FILE',
               default='{0}.log'.format(__package__),
               help='specify name of log file default: %(default)s'),
    cfg.StrOpt('loglevel',
               metavar='LOG_LEVEL',
               default='info',
               help='specify logging level to log messages: %(choices)s',
               choices=['none',
                        'critical',
                        'error',
                        'warning',
                        'info',
                        'debug',
                        'trace']),
    cfg.StrOpt('logconfig',
               metavar='LOG_CONFIG',
               help='specific path and filename of logging configuration \
                    (override defaults)'),
]

cfg.CONF.register_cli_opts(CLI_OPTS)

OPTS = [
    cfg.StrOpt('base_path',
               default=os.getcwd(),
               required=True,
               help='Base path',
               sample_default='/home/username'),
    cfg.StrOpt('base_client_path',
               help='Location torrent client stores data files',
               sample_default='$base_path/torrents/data'),
]

cfg.CONF.register_opts(OPTS)

DB_OPTS = [
    cfg.StrOpt('connection',
               default='sqlite:///$config_dir/torrent.db',
               help='The connection string used to connect to the database'),
    cfg.IntOpt('idle_timeout',
               default=3600,
               help='Timeout before idle sql connections are reaped'),
    cfg.IntOpt('connection_debug',
               default=0,
               help='Verbosity of SQL debugging information. 0=None, 100=All'),
]

cfg.CONF.register_opts(DB_OPTS, 'database')

PROC_OPTS = [
    cfg.IntOpt('max_processes',
               default=4,
               help='max processes to use for performing sync of torrents'),
    cfg.ListOpt('prepare',
                default=[],
                help='name of tasks associated with prepare phase',
                sample_default='filecopy, fileunrar'),
    cfg.ListOpt('activate',
                default=[],
                help='name of tasks associated with activate phase',
                sample_default='filesync'),
    cfg.ListOpt('complete',
                default=[],
                help='name of tasks associated with complete phase',
                sample_default='filedelete'),
]

cfg.CONF.register_opts(PROC_OPTS, group='process')

TASK_OPTS = [
    cfg.StrOpt('sync_path',
               default='/tmp/sync',
               help='Location to temp media copies for syncing to library'),
]

cfg.CONF.register_opts(TASK_OPTS, group='tasks')

SYNC_OPTS = [
    cfg.BoolOpt('dryrun',
                default=False,
                help='rsync dryrun option'),
    cfg.BoolOpt('verbose',
                default=False,
                help='rsync verbose option'),
    cfg.BoolOpt('progress',
                default=False,
                help='rsync progress option'),
    cfg.BoolOpt('perms',
                default=True,
                help='rsync perms option'),
    cfg.BoolOpt('delayupdates',
                default=True,
                help='rsync delayupdates option'),
    cfg.BoolOpt('recursive',
                default=True,
                help='rsync recursive option'),
    cfg.StrOpt('chmod',
               default='ugo+rwx',
               help='rsync chmod option'),
    cfg.StrOpt('identity',
               help='rsync-ssh identity option (ssh key)',
               sample_default='/home/username/.ssh/my_home.key'),
    cfg.StrOpt('port',
               default='22',
               help='rsync-ssh port'),
    cfg.StrOpt('remote_user',
               help='User name on remote system (ssh)',
               sample_default='username'),
    cfg.StrOpt('remote_host',
               help='Host name/IP Address of remote system',
               sample_default='home.example.com'),
    cfg.StrOpt('remote_path',
               help='rsync destination path',
               sample_default='/media/downloads'),
]

cfg.CONF.register_opts(SYNC_OPTS, group='tasks_filesync')

SYNCLOG_OPTS = [
    cfg.StrOpt('stdout_dir',
               default='$config_dir/sync_out',
               help='Output directory for stdout files'),
    cfg.StrOpt('stderr_dir',
               default='$config_dir/sync_err',
               help='Output directory for stderr files'),
    cfg.BoolOpt('stdout_verbose',
                default=False,
                help='Write output to stdout'),
    cfg.BoolOpt('stderr_verbose',
                default=True,
                help='Output verbose details about exceptions'),
]

cfg.CONF.register_opts(SYNCLOG_OPTS, group='tasks_synclog')

TORRENT_OPTS = [
    cfg.StrOpt('torrent_path',
               required=True,
               help='Location of the .torrent files',
               sample_default='/home/username/.config/state'),
    cfg.ListOpt('media_paths',
                required=True,
                help='Location(s) of the media files',
                sample_default='$base_client_path/completed'),
    cfg.StrOpt('incomplete_path',
               required=True,
               help='Location of the downloading torrents',
               sample_default='$base_client_path/inprogress'),
    cfg.ListOpt('video_filetypes',
                default=['.avi', '.mp4', '.mkv', '.mpg'],
                help='List of video filetypes to support. (ignore others)'),
    cfg.ListOpt('compressed_filetypes',
                default=['.rar'],
                help='List of compressed filetypes to support. '
                     '(ignore others)'),
    cfg.IntOpt('minimum_file_size',
               default=75000000,
               help='Minimum file size of a media file'),
]

cfg.CONF.register_opts(TORRENT_OPTS, group='torrent')


def list_opts():
    """Returns a list of oslo_config options available in the library.

    The returned list includes all oslo_config options which may be registered
    at runtime by the library.

    Each element of the list is a tuple. The first element is the name of the
    group under which the list of elements in the second element will be
    registered. A group name of None corresponds to the [DEFAULT] group in
    config files.

    The purpose of this is to allow tools like the Oslo sample config file
    generator to discover the options exposed to users by this library.

    :returns: a list of (group_name, opts) tuples
    """
    from seedbox.common import tools
    all_opts = []
    all_opts.extend(tools.make_opt_list([OPTS], None))
    all_opts.extend(tools.make_opt_list([DB_OPTS], 'database'))
    all_opts.extend(tools.make_opt_list([PROC_OPTS], 'process'))
    all_opts.extend(tools.make_opt_list([TASK_OPTS], 'tasks'))
    all_opts.extend(tools.make_opt_list([SYNC_OPTS], 'tasks_filesync'))
    all_opts.extend(tools.make_opt_list([SYNCLOG_OPTS], 'tasks_synclog'))
    all_opts.extend(tools.make_opt_list([TORRENT_OPTS], 'torrent'))
    return all_opts
